### ISSUE TITLE
Typescript (and preprocessing) support?

### ISSUE THREAD

**Issue by jadbox** [jakedeichert/svelvet#24]
How would I set up typescript support with a new project using this?

**Comment by jakedeichert** [comment]
Great question! Svelte itself has an issue for [typescript support](https://github.com/sveltejs/svelte/issues/1639). One of the linked solutions is [svelte-preprocess](https://github.com/kaisermann/svelte-preprocess) which is a community effort to get it working. I'm sure you may already be aware of these efforts.

As for svelvet, i know it doesn't work with typescript yet because i don't plug-into the svelte compiler to do any preprocessing at all. For the time being, i don't see this changing unfortunately! If someone figures out a zero-config way to get this to work, that would be cool... but i'd much rather way for first-party typescript support in svelte itself :)

**Comment by v0idifier** [comment]
wouldn't a custom `svelte.config.js` allow it?

**Comment by jakedeichert** [comment]
As far as i know, svelte doesn't actually load any `svelte.config.js` yet 🤔 
Or does `svelte-preprocess` look for a `svelte.config.js` file?

I've only skimmed their docs, so i'm not very familiar. It looks like we'd have to plug svelte-preprocess into [svelte's preprocessing api](https://svelte.dev/docs#svelte_preprocess).

**Comment by v0idifier** [comment]
what I mean is that I think having a `svelte.config.js` read out by svelvet and passing it to svelte compile configs would allow the usage of the `preprocess` option, allowing the use of tools like `svelte-preprocess` without special handling on svelvet's side

**Comment by v0idifier** [comment]
nevermind, svelte doesn't have a `preprocess` option, it's handled specially by `svelte-loader` (see https://github.com/sveltejs/svelte-loader/blob/master/index.js)

**Comment by jakedeichert** [comment]
[Svelte has a preprocessing api](https://svelte.dev/docs#svelte_preprocess) though. I'm not sure if there's a zero-config way to support all possible preprocessing scenarios though.

Maybe a flag that points to an installed module name to use as the preprocessor.
```sh
svelvet --preprocess svelte-preprocess 
# or 
svelvet --preprocess ./scripts/my-custom-preprocessor
# a custom preprocessor could wrap svelte-preprocess to pass config to it, for example.
```

With that said, i don't want to jump to this solution today. I'd rather lock us into **not** supporting preprocessing (in any form) for 3 months at least, and see how people use svelvet and what issues they run into. I'd love to know a percentage estimate of how many people in the svelte community actually preprocess their code.

**Comment by v0idifier** [comment]
preprocessing can and also is used for `sass`, `stylus` or anything else support, and for `pug` or other templating engines instead of normal html

**Comment by jakedeichert** [comment]
Yupp, i definitely see value in that.. still would be nice to know the percentage of the community that really depends on those!

I know we'll need to add preprocessing support at some point.

The two options we have so far are: 

1. Look for a `svelte.config.js` file and let people supply custom compiler and preprocessing config.
2. Add a flag to specify a preprocessor (`svelvet --preprocess svelte-preprocess`)


I kind of like option 2 because `svelte.config.js` isn't a thing and feels a bit wrong to start using that filename. Really still want to avoid the need for our own file `svelvet.config.js`.

**Comment by v0idifier** [comment]
keep in mind that some people may want to use multiple preprocess and you can pass options as an object when using `svelte-preprocess`, so it isn't a good option either. maybe using fields from `package.json`

**Comment by jakedeichert** [comment]
Yeah, that's kind of why i suggested this:

```
svelvet --preprocess ./scripts/my-custom-preprocessor
# a custom preprocessor could wrap svelte-preprocess to pass config to it, for example.
```

That way people can configure whatever preprocessor(s) they want and we just pass that to svelte. The advantage here is that we don't need to document or depend on any specific config file 👍

**Comment by v0idifier** [comment]
sounds good, however wouldn't that create too much file bloat?

**Comment by jakedeichert** [comment]
Hmm, in which way?

If you're talking about package.json scripts line length due to the flag, yeah it's a bit longer but still avoids the need for an actual config file. I also think we would not support multiple preprocessor flags, but your custom script could combine them appropriately.

If we can't support this in a powerful enough way, that might be a sign that people may require a more complex bundler like webpack or rollup.

**Comment by v0idifier** [comment]
personally i think svelvet has the role of a build process, where bundlers like webpack or rollup don't only bundle but also act as that build process. i thought that for each preprocessor a file would be needed, but if it's all in one file there won't be `n` files per preprocessor

**Comment by jakedeichert** [comment]
> i thought that for each preprocessor a file would be needed, but if it's all in one file there won't be n files per preprocessor (so that's good)

Exactly, just one script should be needed to combine all preprocessors you need.. and that'll be on the user.

> personally i think svelvet has the role of a build process
Agreed. Preprocessing is a core component of svelte of course, and we need to try to support that 👍

**Comment by theword** [comment]
First I just want to say thank you for doing this project. And while I don't know if svelte and snowpack will be the future winners, I think the ideas behinds these are the rights ideas. So thank you for pushing the envelope.  

I was looking to add pug to a new project of mine just because I like the look of it better than HTML. But based on this thread, I would have to sacrifice using snowpack? 

And on a side note, I haven't gone down this road yet but will be soon, was wondering with your greater knowledge of the subject, whether you foresee there being any difficulties using this for a project in AWS? I had tried to use rollup in an AWS project but it wouldn't take it so I had to go back to webpack. However, since this works in a completely different way, I don't know if it's necessary for AWS to support it or whether this should just work using the web modules?

**Comment by jakedeichert** [comment]
> I was looking to add pug to a new project of mine just because I like the look of it better than HTML. But based on this thread, I would have to sacrifice using snowpack?

Do you use pug with svelte-preprocess already? It looks like they have pug support at least. Anyway, you do not need to sacrifice using snowpack. This issue is about adding support for preprocessing into svelvet itself, and that would still work with snowpack once it's integrated. You just can't use pug (nor any preprocessor) with svelvet as of today.


> whether you foresee there being any difficulties using this for a project in AWS

I don't foresee any issues... but i guess it depends what you're talking about. Svelvet outputs a static `dist` directory which contains all your website assets + js scripts in esm format. You'd just have to host that directory somewhere like s3 and it should work just like any other static website.

**Comment by theword** [comment]
No, I haven't used pug with svelte pre-process yet. I've only used pug on a rails project. However, you're saying that I won't be able to use it with svelvet.

Thanks, that's good to know, I was referring to AWS Ampify and Firebase type applications in particular but yeah, you're right, shouldn't be any different from any other static site. 

Also one other random side question. On the Snowpack tutorial, they recommend using servor instead of serve because it has live reload. 

Tutorial (Serve & Run Your Application):
https://www.snowpack.dev/#installing-a-dev-server

The NPM Servor Package:
https://www.npmjs.com/package/servor

However, when I run their command npx servor --reload (or npx servor --reload dist)
It shows up as a blank page, have you tried running svelvet with servor? I'm probably messing up the command. But it does seem to be an easy way for you to add auto-refresh to the project by using this other dev server with minimum effort.

**Comment by jadbox** [comment]
@theword might be better to move this conversation to another Issue [serving static files vs typescript suport]

**Comment by cdock1029** [comment]
Pre-processing would be great to easily integrate tailwindcss and purgecss as described here:

https://dev.to/sarioglu/using-svelte-with-tailwindcss-a-better-approach-47ph

**Comment by jakedeichert** [comment]
Just released v0.3.4 with preprocessor configuration support :)

Usage info here: https://github.com/jakedeichert/svelvet#can-i-preprocess-my-svelte-components

Full example with svelte-preprocess here: https://github.com/jakedeichert/svelvet/tree/master/tests/snapshot-preprocessors
